File: sonar-server/src/test/java/org/sonar/server/ws/RequestTest.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2013 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.server.ws;

import com.google.common.collect.ImmutableMap;
import org.junit.Test;

import static org.fest.assertions.Assertions.assertThat;

public class RequestTest {
  @Test
  public void string_params() {
    Request request = new Request(ImmutableMap.of("foo", "bar"));
    assertThat(request.param("none")).isNull();
    assertThat(request.param("foo")).isEqualTo("bar");
  }

  @Test
  public void int_params() {
    Request request = new Request(ImmutableMap.of("foo", "123"));
    assertThat(request.intParam("none")).isNull();
    assertThat(request.intParam("foo")).isEqualTo(123);

    assertThat(request.intParam("none", 456)).isEqualTo(456);
    assertThat(request.intParam("foo", 456)).isEqualTo(123);
  }
}

File: server/sonar-web/src/main/js/sonar-aligned/helpers/__tests__/request-test.ts
/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { setImmediate } from 'timers';
import { HttpStatus } from '../../../helpers/request';
import { Dict } from '../../../types/types';
import { getJSON } from '../request';

const url = '/my-url';

beforeEach(() => {
  jest.clearAllMocks();
  window.fetch = jest.fn().mockResolvedValue(mockResponse({}, HttpStatus.Ok, {}));
});

describe('getJSON', () => {
  it('should get json without parameters', async () => {
    const response = mockResponse({}, HttpStatus.Ok, {});
    window.fetch = jest.fn().mockResolvedValue(response);
    getJSON(url);
    await new Promise(setImmediate);

    expect(window.fetch).toHaveBeenCalledWith(url, expect.objectContaining({ method: 'GET' }));
    expect(response.json).toHaveBeenCalled();
  });

  it('should get json with parameters', () => {
    getJSON(url, { data: 'test' });
    expect(window.fetch).toHaveBeenCalledWith(
      url + '?data=test',
      expect.objectContaining({ method: 'GET' }),
    );
  });
});

function mockResponse(headers: Dict<string> = {}, status = HttpStatus.Ok, value?: any): Response {
  const body = value && value instanceof Object ? JSON.stringify(value) : value;
  const response = new Response(body, { headers, status });
  response.json = jest.fn().mockResolvedValue(value);
  response.text = jest.fn().mockResolvedValue(value);
  return response;
}
